File: 백준/Silver/1393. 음하철도 구구팔/음하철도 구구팔.cc
/**
 * Author: nowalex322, Kim HyeonJae
 */
#include <bits/stdc++.h>
using namespace std;

// #define int long long
#define MOD 1000000007
#define INF LLONG_MAX
#define ALL(v) v.begin(), v.end()

#ifdef LOCAL
#include "algo/debug.h"
#else
#define debug(...) 42
#endif

void solve() {
    int x0, y0;
    cin >> x0 >> y0;

    int x1, y1, dx, dy;
    cin >> x1 >> y1 >> dx >> dy;

    int g = gcd(abs(dx), abs(dy));
    if (g > 0) {
        dx /= g;
        dy /= g;
    }

    long long min_dist = LLONG_MAX;
    int tx = x1, ty = y1;

    for (int t = 0; t < 200; t++) {
        int currX = x1 + t * dx;
        int currY = y1 + t * dy;
        long long dist = 1LL * (x0 - currX) * (x0 - currX) +
                         1LL * (y0 - currY) * (y0 - currY);

        if (dist < min_dist) {
            min_dist = dist;
            tx = currX;
            ty = currY;
        }
    }

    cout << tx << " " << ty << "\n";
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int tt = 1;  // 기본적으로 1번의 테스트 케이스를 처리
    // cin >> tt;    // 테스트 케이스 수 입력 (필요 시)

    while (tt--) {
        solve();
    }
    return 0;
}

File: 백준/Silver/1393. 음하철도 구구팔/음하철도 구구팔.java
/**
 * Author: nowalex322, Kim HyeonJae
 */
import java.io.*;
import java.util.*;

public class Main {
	static BufferedReader br;
	static BufferedWriter bw;
	static StringTokenizer st;

	public static void main(String[] args) throws Exception {
		new Main().solution();
	}

	public void solution() throws Exception {
		br = new BufferedReader(new InputStreamReader(System.in));
//		br = new BufferedReader(new InputStreamReader(new FileInputStream("input.txt")));
		bw = new BufferedWriter(new OutputStreamWriter(System.out));
		st = new StringTokenizer(br.readLine());
		int x0 = Integer.parseInt(st.nextToken());
		int y0 = Integer.parseInt(st.nextToken());
		st = new StringTokenizer(br.readLine());
		int x1 = Integer.parseInt(st.nextToken());
		int y1 = Integer.parseInt(st.nextToken());
		int dx = Integer.parseInt(st.nextToken());
		int dy = Integer.parseInt(st.nextToken());
		
		int gcd = gcd(Math.abs(dx), Math.abs(dy));
	    if (gcd > 0) {
	        dx /= gcd;
	        dy /= gcd;
	    }
	    
	    long min = Long.MAX_VALUE;
	    int tx = x1, ty = y1;
	    
	    for (int t = 0; t < 200; t++) {
	        int currX = x1 + t * dx;
	        int currY = y1 + t * dy;
	        long dist = (long)(x0 - currX) * (x0 - currX) + (long)(y0 - currY) * (y0 - currY);
	        
	        if (dist < min) {
	            min = dist;
	            tx = currX;
	            ty = currY;
	        }
	    }
	    
	    bw.write(tx + " " + ty);
		bw.flush();
		bw.close();
		br.close();
	}

	private int gcd(int a, int b) {
		while (b != 0) {
	        int tmp = a % b;
	        a = b;
	        b = tmp;
	    }
	    return a;
	}
}